### ISSUE TITLE
Bridge promises don't handle multiple resolve or reject calls properly

### ISSUE THREAD

**Issue by esprehn** [facebook/react-native#20262]
- [ x] Review the documentation: https://facebook.github.io/react-native
  - [ x] Search for existing issues: https://github.com/facebook/react-native/issues
  - [ x] Use the latest React Native release: https://github.com/facebook/react-native/releases

## Environment

Environment:
  OS: macOS High Sierra 10.13.5
  Node: 8.9.4
  Yarn: 1.5.1
  npm: 5.6.0
  Watchman: 4.7.0
  Xcode: Xcode 9.4.1 Build version 9F2000
  Android Studio: 3.1 AI-173.4720617

Packages: (wanted => installed)
  react: 16.2.0 => 16.2.0
  react-native: 0.54.4 => 0.54.4

## Description
PromiseImpl and RCTModuleMethod don't handle promise resolution properly. Resolving a promise a second time is supposed to have no effect per ES semantics, similarly rejecting it a second time, or if it's rejected already, should have no effect.

PromiseImpl.resolve a second time will throw an error:

```
Exception java.lang.RuntimeException: Illegal callback invocation from native module. This callback type only permits a single invocation from native code.
com.facebook.react.bridge.CallbackImpl.invoke (CallbackImpl.java:30)
com.facebook.react.bridge.PromiseImpl.resolve (PromiseImpl.java:32)
```

For iOS this doesn't crash in ObjC, instead it sends another message over the bridge, but the callbacks are gone, so it incorrectly warns in a dev build (`Callback with id ${cbID}: ${module}.${method}() not found`) and is correctly a no-op in a release build (though it is inefficient to send the bridge response).

## Reproducible Demo

Create a `@RectMethod` and do resolve() in the Java twice, or do do resolve twice in an ObjC iOS method.

**Comment by esprehn** [comment]
This happens in the most recent version too. The code is the same.

**Comment by hramos** [comment]
@esprehn can you update your post accordingly? 0.54.4 is not the latest as of this writing, and we'd like to know the exact version you tested in.

**Comment by esprehn** [comment]
This does actually crash on iOS in a debug build, I misread the code:
https://github.com/facebook/react-native/blob/30b9d81087cb86f5fb272d00bfee63a0632009f5/React/Base/RCTModuleMethod.mm#L119

**Comment by esprehn** [comment]
This bug still exists. @hramos The bot probably shouldn't close issues with pending PRs?

**Comment by hramos** [comment]
@esprehn the bot won't close issues with a pending PR, but this issue was not labeled as such at the time of closing. Can you update your original post and make sure it mentions the actual version you tested in? Otherwise the bot might close it again for referring to an older version (0.56 is latest at this time).

**Comment by esprehn** [comment]
This issue still exists. Bridge promises don't follow the spec.

**Comment by dulmandakh** [comment]
This issue if fixed in 0.59. Thank you
